### PR TITLE
Flight termination: terminate if failure is detected on takeoff

### PR DESCRIPTION
**Pull request by bresch**
During the first few seconds after takeoff, the failure detector is allowed to trigger motor lockdown.
This is done for safety reasons to detect tipping-over or unstable tuning gains.

SITL test: takeoff with unstable gains
https://logs.px4.io/plot_app?log=b43c05ec-0330-414a-b069-41309ce13cf5

FYI @Jaeyoung-Lim

### CODE REVIEW

**Comment by bresch** [comment]
@hamishwillee The documentation for flight termination is [here](https://docs.px4.io/master/en/advanced_config/flight_termination.html) but where exactly should we mention that the failure detector is active during the first few seconds after takeoff even if the circuit-breaker is set? Should it be just a "note" and a modification of the diagram?

**Comment by hamishwillee** [comment]
Hi @bresch 
- Is the whole failure detector enabled on takeoff or just the attitude trigger? i.e. could an ATS still trigger on takeoff with the circuit breaker enabled?
- Is the failure detector enabled before arming, or after arming before takeoff?

In any case, first I'd suggest we update the text in [Safety > Failure Detector](https://docs.px4.io/master/en/config/safety.html#failure_detector). Perhaps replace the existing note with.

> **Note** Failure detection is always enabled for the first few seconds after takeoff. It can be enabled/disabled in flight using the [CBRK_FLIGHTTERM=0](#CBRK_FLIGHTTERM) circuit breaker.

For the [flight termination doc](https://docs.px4.io/master/en/advanced_config/flight_termination.html) I would change the first pullet point under the software configuration heading to:

- Enable the failure detector during flight by setting `CBRK_FLIGHTTERM=0` (disabled by default).
  > **Note** The failure detector active for a few seconds after takeoff, irrespective of the value of `CBRK_FLIGHTTERM` (in order to detect the vehicle tipping during launch).

Updating diagram would be good. Perhaps you could just add a text where shown saying something like "(After takeoff complete)" where marked?

image

**Comment by bkueng** [comment]
So it would trigger a parachute?

**Comment by hamishwillee** [comment]
@bkueng Good question. I would assume so. @bresch ?

**Comment by bresch** [comment]
@bkueng It triggers flight termination, so if you have flight termination associated with a parachute trigger, yes. Should we trigger a "manual lockdown" (aka: kill) ?
I thought this wasn't a big issue because most of the new setups that would benefit from this feature won't have a parachute for their first flight, and even if it triggers it, you would have this as well with an external flight termination system. However, I agree that a simple kill would be enough because the parachute will most of the times not help, except if the drone shoots in the sky fast enough.

**Comment by julianoes** [comment]
I'm scared that this will lead to false positives where we trigger termination after having taken off successfully within 5 seconds. Which detectors are used here and how sensitive are they?

**Comment by bresch** [comment]
@julianoes By default, it's the tilt limit of 60 degrees (defined by [FD_FAIL_P](https://dev.px4.io/master/en/advanced/parameter_reference.html#FD_FAIL_P)).  We can reduce the 5s period or increase the default if required, but I've never seen a false detection in any of our flights

**Comment by julianoes** [comment]
Ok, which means this won't work when flying acro, and maybe planes.

**Comment by bresch** [comment]
@julianoes Correct, the failure detector is not active in acro, rattitude and FW manual modes

**Comment by LorenzMeier** [comment]
@bresch I don't think termination is the right choice here. I would limit to 3 seconds or less and lock the system down rather than triggering termination.

Otherwise we will have parachute deployments on the ground and some of them are kinetic (explosives).

File: src/modules/commander/Commander.cpp [hunk 1]
```diff
@@ -758,13 +758,19 @@ Commander::handle_command(vehicle_status_s *status_local, const vehicle_command_
...
+				}
 
 			} else if (cmd.param1 > 0.5f) {
 				//XXX update state machine?
-				armed_local->force_failsafe = true;
-				warnx("forcing failsafe (termination)");
+				if (!_flight_termination_triggered) {
+					armed_local->force_failsafe = true;
+					_flight_termination_triggered = true;
+					warnx("forcing failsafe (termination)");
```
Comment (LorenzMeier) [inline src/modules/commander/Commander.cpp]: Should this be PX4_WARN()?

Comment (LorenzMeier) [reply src/modules/commander/Commander.cpp]: Or be removed?

Comment (bresch) [reply src/modules/commander/Commander.cpp]: Right, I think warnx is deprecated, will change that

**Comment by hamishwillee** [comment]
@bresch Did you make the changes needed in docs for this? i.e. as discussed in https://github.com/PX4/Firmware/pull/14428#issuecomment-601510447

Did this end up being "flight termination" or some other action? Still triggered by the failure detector?

**Pull request merged**